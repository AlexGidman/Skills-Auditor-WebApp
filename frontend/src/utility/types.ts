export const SYSTEM_ROLES = ["StaffUser", "Manager", "Admin"];
export const STAFF_USER = SYSTEM_ROLES[0];
export const MANAGER_SR = SYSTEM_ROLES[1];
export const ADMIN = SYSTEM_ROLES[2];

export const JOB_ROLES = ["Mid-Level Developer", "Senior Developer", "Manager"];
export const MIDLEVEL_DEVELOPER = JOB_ROLES[0];
export const SENIOR_DEVELOPER = JOB_ROLES[1];
export const MANAGER_JR = JOB_ROLES[2];

export const TOKEN_COOKIE = "token";

// Week in seconds
export const TOKEN_LIFE = 60 * 60 * 24 * 7;

export const SKILL_LEVELS = ["None", "Basic", "Intermediate", "High", "Expert"];
export const NONE = SKILL_LEVELS[0];
export const BASIC = SKILL_LEVELS[1];
export const INTERMEDIATE = SKILL_LEVELS[2];
export const HIGH = SKILL_LEVELS[3];
export const EXPERT = SKILL_LEVELS[4];

export type User = {
    firstName: string;
    lastName: string;
    email: string;
    password: string;
};
