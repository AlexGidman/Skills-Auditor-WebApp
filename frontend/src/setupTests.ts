// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import * as apiRequests from "./utility/apiRequests";
import * as reactCookie from "react-cookie";

import { MANAGER_JR, ADMIN, User } from "./utility/types";

jest.mock("uuid", () => {
    return {
        ...jest.requireActual("uuid"),
        v4: jest.fn().mockReturnValue("uuid"),
    };
});

jest.mock("react-cookie", () => {
    return {
        ...jest.requireActual("react-cookie"),
        useCookies: jest.fn(),
    };
});

export const mockApiRequests = {
    getAllCategories: jest.spyOn(apiRequests, "getAllCategories"),
    getCategory: jest.spyOn(apiRequests, "getCategory"),
    updateCategory: jest.spyOn(apiRequests, "updateCategory"),
    getUserDetails: jest.spyOn(apiRequests, "getUserDetails"),
    updateUserDetails: jest.spyOn(apiRequests, "updateUserDetails"),
    getSkill: jest.spyOn(apiRequests, "getSkill"),
    updateSkill: jest.spyOn(apiRequests, "updateSkill"),
    getAllSkills: jest.spyOn(apiRequests, "getAllSkills"),
    addUser: jest.spyOn(apiRequests, "addUser"),
    addDirectReport: jest.spyOn(apiRequests, "addDirectReport"),
    getUserFromToken: jest.spyOn(apiRequests, "getUserFromToken"),
    getAllDirectReports: jest.spyOn(apiRequests, "getAllDirectReports"),
    getAllStaffSkills: jest.spyOn(apiRequests, "getAllStaffSkills"),
    deleteUser: jest.spyOn(apiRequests, "deleteUser"),
    getStaffSkill: jest.spyOn(apiRequests, "getStaffSkill"),
    updateStaffSkill: jest.spyOn(apiRequests, "updateStaffSkill"),
    postLogin: jest.spyOn(apiRequests, "postLogin"),
    addSkill: jest.spyOn(apiRequests, "addSkill"),
    addCategory: jest.spyOn(apiRequests, "addCategory"),
    deleteStaffSkill: jest.spyOn(apiRequests, "deleteStaffSkill"),
    addUserSkill: jest.spyOn(apiRequests, "addUserSkill"),
};

export const mockAdminUser = {
    id: "1",
    firstName: "Admin",
    lastName: "User",
    email: "Admin@email.com",
    jobRole: MANAGER_JR,
    systemRole: ADMIN,
};

export const mockError = { response: { data: { error: new Error("Oops") }, status: 400 } };

beforeAll(() => {
    console.log = jest.fn();
    console.error = jest.fn();
});

beforeEach(() => {
    jest.spyOn(reactCookie, "useCookies").mockImplementation(() => [
        { token: "Bearer token" },
        jest.fn(),
        jest.fn(),
    ]);
    // @ts-ignore
    mockApiRequests.getUserFromToken.mockImplementation(() =>
        Promise.resolve({ data: mockAdminUser }),
    );
});

afterEach(() => {
    jest.resetAllMocks();
});
