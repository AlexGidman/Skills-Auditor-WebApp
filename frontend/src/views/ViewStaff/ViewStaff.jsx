import React, { useEffect, useState } from "react";
import { useOutletContext, useNavigate } from "react-router-dom";
import { LoadingPlaceholder } from "../../components/LoadingPlaceholder/LoadingPlaceholder";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { LinkButton } from "../../components/LinkButton/LinkButton";
import { Button } from "../../components/Button/Button";

import { ListView } from "../../components/ListView/ListView";
import {
    useAPI,
    useIsAdminOrManager,
    getSelectOptionsFromArray,
    formatSkillLevel,
} from "../../utility/helper";
import {
    getAllDirectReports,
    getAllStaffSkills,
    deleteUser,
    deleteStaffSkill,
} from "../../utility/apiRequests";
import styles from "./ViewStaff.module.css";

export const ViewStaff = () => {
    useIsAdminOrManager();
    const [currentUser] = useOutletContext();

    const { data, loading, error } = useAPI(getAllDirectReports, [currentUser.id]);
    if (loading) return <LoadingPlaceholder />;
    if (error) return <ErrorMessage error={error} />;

    const sortedStaff = data.sort((a, b) => a.report.forename.localeCompare(b.report.forename));

    return (
        <>
            <header>
                <h1>View Staff</h1>
            </header>
            <section>
                <Form data={sortedStaff} />
            </section>
        </>
    );
};

const Form = ({ data }) => {
    const [selectedStaff, setSelectedStaff] = useState(data[0]);
    const [staffSkills, setStaffSkills] = useState([]);
    const [selectedStaffSkill, setSelectedStaffSkill] = useState();
    const [refresh, setRefresh] = useState(false);
    const [, setShowToast] = useOutletContext();
    const navigate = useNavigate();

    const {
        data: staffSkillData,
        loading: staffSkillsLoading,
        error: staffSkillsError,
        callback: getSkillData,
    } = useAPI(getAllStaffSkills, [selectedStaff.report.id], true);

    const {
        data: deleteStaffSkillData,
        loading: deleteStaffSkillLoading,
        error: deleteStaffSkillError,
        callback: handleDeleteStaffSkill,
    } = useAPI(deleteStaffSkill, [selectedStaffSkill ? selectedStaffSkill.id : ""], true);

    const {
        data: deleteUserData,
        loading: deleteUserLoading,
        error: deleteUserError,
        callback: handleDeleteUser,
    } = useAPI(deleteUser, [selectedStaff.report.id], true);

    useEffect(() => {
        selectedStaff && getSkillData();
    }, [selectedStaff]);

    useEffect(() => {
        staffSkillData && setStaffSkills(staffSkillData);
    }, [staffSkillData]);

    useEffect(() => {
        staffSkills && setSelectedStaffSkill(staffSkills[0]);
    }, [staffSkills]);

    useEffect(() => {
        if (refresh && !deleteUserLoading && !deleteStaffSkillLoading) {
            setRefresh(false);
            deleteUserError && setShowToast({ error: "Failed to delete user" });
            deleteStaffSkillError && setShowToast({ error: "Failed to delete staff skill" });
            (deleteUserData || deleteStaffSkillData) && navigate(0);
        }
    }, [
        deleteStaffSkillData,
        deleteStaffSkillLoading,
        deleteStaffSkillError,
        deleteUserData,
        deleteUserError,
        deleteUserLoading,
        navigate,
        refresh,
        setShowToast,
    ]);

    return (
        <form className={styles.form}>
            <ListView
                className={styles.listview}
                options={getSelectOptionsFromArray(
                    data.map((staff) => {
                        return staff.report.id;
                    }),
                    data.map((staff) => {
                        return `${staff.report.forename} ${staff.report.surname}`;
                    }),
                )}
                value={selectedStaff.report.id}
                labelText="Direct Reports"
                onChange={(e) => {
                    setSelectedStaff(
                        data.filter((staff) => staff.report.id.toString() === e.target.value)[0],
                    );
                    setStaffSkills([]);
                }}
            />
            <div className={styles.flexbox}>
                <table className={styles.table}>
                    <tbody>
                        <tr>
                            <th>Email</th>
                            <td>{selectedStaff.report.email}</td>
                        </tr>
                        <tr>
                            <th>Job role</th>
                            <td>{selectedStaff.report.job_role}</td>
                        </tr>
                        <tr>
                            <th>System role</th>
                            <td>{selectedStaff.report.system_role}</td>
                        </tr>
                    </tbody>
                </table>
                {staffSkillsLoading ? (
                    <LoadingPlaceholder />
                ) : staffSkills ? (
                    <ListView
                        className={styles.skillsList}
                        options={
                            staffSkills
                                ? getSelectOptionsFromArray(
                                      staffSkills.map((skill) => {
                                          return skill.id;
                                      }),
                                      staffSkills.map((skill) => {
                                          return `${skill.skillID.name} - ${formatSkillLevel(
                                              skill.skill_level,
                                          )}`;
                                      }),
                                  )
                                : null
                        }
                        value={selectedStaffSkill ? selectedStaffSkill.id : null}
                        labelText="Skills"
                        onChange={(e) => {
                            setSelectedStaffSkill(
                                staffSkills.filter((skill) => {
                                    return skill.id.toString() === e.target.value;
                                })[0],
                            );
                        }}
                    />
                ) : staffSkillsError ? (
                    <ErrorMessage error={staffSkillsError} />
                ) : null}
            </div>
            <div className={styles.flexbox}>
                <LinkButton
                    className={styles.button}
                    name="Edit User"
                    disabled={!selectedStaff}
                    to={`/edit/user/${selectedStaff?.report?.id || ""}`}
                />
                <Button
                    className={styles.button}
                    disabled={!selectedStaff}
                    onClick={(e) => {
                        e.preventDefault();
                        handleDeleteUser();
                        setRefresh(true);
                    }}
                >
                    Delete Report
                </Button>
                <LinkButton
                    className={styles.button}
                    name="Add Skill"
                    to={`/addstaffskill/${selectedStaff?.report?.id || ""}`}
                />
                <LinkButton
                    className={styles.button}
                    name="Edit Skill"
                    disabled={!selectedStaffSkill}
                    to={`/edit/staffskill/${selectedStaffSkill?.id || ""}`}
                />
                <Button
                    className={styles.button}
                    disabled={!selectedStaffSkill}
                    onClick={(e) => {
                        e.preventDefault();
                        handleDeleteStaffSkill();
                        setRefresh(true);
                    }}
                >
                    Delete Skill
                </Button>
            </div>
        </form>
    );
};
