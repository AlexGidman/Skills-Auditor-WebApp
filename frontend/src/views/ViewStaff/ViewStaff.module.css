.button {
    flex: 20%;
    min-width: 16rem;
    box-sizing: border-box;
}
.flexbox {
    display: flex;
    justify-content: space-between;
    gap: var(--s2);
    flex-wrap: wrap;
    margin-bottom: var(--s2);
}

.form {
    margin: 0 10%;
}

.listview {
    margin-bottom: var(--s2);
}

.skillsList {
    composes: listview;

    flex: 30%;
}

.table {
    flex: 40%;
}

@media (max-width: 900px) {
    th,
    td {
        font-size: small;
    }
    .form {
        margin: 0;
    }
}
