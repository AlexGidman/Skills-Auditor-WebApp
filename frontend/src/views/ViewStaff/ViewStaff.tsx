import React, { useEffect, useState } from "react";
import { useOutletContext, useNavigate } from "react-router-dom";
import { LoadingPlaceholder } from "../../components/LoadingPlaceholder/LoadingPlaceholder";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { LinkButton } from "../../components/LinkButton/LinkButton";
import { Button } from "../../components/Button/Button";

import { ListView } from "../../components/ListView/ListView";
import {
    useAPI,
    useIsAdminOrManager,
    getSelectOptionsFromArray,
    formatSkillLevel,
} from "../../utility/helper";
import {
    getAllDirectReports,
    getAllStaffSkills,
    deleteUser,
    deleteStaffSkill,
} from "../../utility/apiRequests";
import styles from "./ViewStaff.module.css";
import { DirectReport, StaffSkill } from "../../utility/types";
import { AppOutletContext } from "../AppWrapper/AppWrapper";

export const ViewStaff = () => {
    useIsAdminOrManager();
    const { currentUser } = useOutletContext<AppOutletContext>();

    const { data, loading, error } = useAPI<DirectReport[]>(getAllDirectReports, [currentUser.id]);
    if (loading) return <LoadingPlaceholder />;
    if (error) return <ErrorMessage error={error} />;
    if (data) {
        const sortedStaff = data.sort((a, b) =>
            a.report.firstName.localeCompare(b.report.firstName),
        );
        return (
            <>
                <header>
                    <h1>View Staff</h1>
                </header>
                <section>
                    <Form data={sortedStaff} />
                </section>
            </>
        );
    }
    return null;
};

interface FormProps {
    data: DirectReport[];
}

const Form = ({ data }: FormProps) => {
    const [selectedStaff, setSelectedStaff] = useState(data[0]);
    const [staffSkills, setStaffSkills] = useState<StaffSkill[]>([]);
    const [selectedStaffSkill, setSelectedStaffSkill] = useState<StaffSkill>();

    const [refresh, setRefresh] = useState(false);
    const { setShowToast } = useOutletContext<AppOutletContext>();
    const navigate = useNavigate();

    const {
        data: staffSkillData,
        loading: staffSkillsLoading,
        error: staffSkillsError,
        callback: getSkillData,
    } = useAPI<StaffSkill[]>(getAllStaffSkills, [selectedStaff.report.id], true);

    const {
        data: deleteStaffSkillData,
        loading: deleteStaffSkillLoading,
        error: deleteStaffSkillError,
        callback: handleDeleteStaffSkill,
    } = useAPI(deleteStaffSkill, [selectedStaffSkill ? selectedStaffSkill.id : ""], true);

    const {
        data: deleteUserData,
        loading: deleteUserLoading,
        error: deleteUserError,
        callback: handleDeleteUser,
    } = useAPI(deleteUser, [selectedStaff.report.id], true);

    useEffect(() => {
        selectedStaff && getSkillData();
    }, [selectedStaff]);

    useEffect(() => {
        staffSkillData && setStaffSkills(staffSkillData);
    }, [staffSkillData]);

    useEffect(() => {
        staffSkills && setSelectedStaffSkill(staffSkills[0]);
    }, [staffSkills]);

    useEffect(() => {
        if (refresh && !deleteUserLoading && !deleteStaffSkillLoading) {
            setRefresh(false);
            deleteUserError && setShowToast({ error: "Failed to delete user" });
            deleteStaffSkillError && setShowToast({ error: "Failed to delete staff skill" });
            (deleteUserData || deleteStaffSkillData) && navigate(0);
        }
    }, [
        deleteStaffSkillData,
        deleteStaffSkillLoading,
        deleteStaffSkillError,
        deleteUserData,
        deleteUserError,
        deleteUserLoading,
        navigate,
        refresh,
        setShowToast,
    ]);

    return (
        <form className={styles.form}>
            <ListView
                className={styles.listview}
                options={getSelectOptionsFromArray(
                    data.map((staff) => {
                        return staff?.report?.id?.toString() || "";
                    }),
                    data.map((staff) => {
                        return `${staff.report.firstName} ${staff.report.lastName}`;
                    }),
                )}
                value={selectedStaff.report.id}
                labelText="Direct Reports"
                onChange={(e) => {
                    setSelectedStaff(
                        data.filter((staff) => staff?.report?.id?.toString() === e.target.value)[0],
                    );
                    setStaffSkills([]);
                }}
            />
            <div className={styles.flexbox}>
                <table className={styles.table}>
                    <tbody>
                        <tr>
                            <th>Email</th>
                            <td>{selectedStaff.report.email}</td>
                        </tr>
                        <tr>
                            <th>Job role</th>
                            <td>{selectedStaff.report.jobRole}</td>
                        </tr>
                        <tr>
                            <th>System role</th>
                            <td>{selectedStaff.report.systemRole}</td>
                        </tr>
                    </tbody>
                </table>
                {staffSkillsLoading ? (
                    <LoadingPlaceholder />
                ) : staffSkills ? (
                    <ListView
                        className={styles.skillsList}
                        options={
                            staffSkills
                                ? getSelectOptionsFromArray(
                                      staffSkills.map((staffSkill) => {
                                          return staffSkill.id;
                                      }),
                                      staffSkills.map((staffSkill) => {
                                          return `${staffSkill.skill.name} - ${formatSkillLevel(
                                              staffSkill.skillLevel,
                                          )}`;
                                      }),
                                  )
                                : []
                        }
                        value={selectedStaffSkill ? selectedStaffSkill.id : ""}
                        labelText="Skills"
                        onChange={(e) => {
                            setSelectedStaffSkill(
                                staffSkills.filter((skill) => {
                                    return skill.id.toString() === e.target.value;
                                })[0],
                            );
                        }}
                    />
                ) : staffSkillsError ? (
                    <ErrorMessage error={staffSkillsError} />
                ) : null}
            </div>
            <div className={styles.flexbox}>
                <LinkButton
                    className={styles.button}
                    name="Edit User"
                    disabled={!selectedStaff}
                    path={`/edit/user/${selectedStaff?.report?.id || ""}`}
                />
                <Button
                    className={styles.button}
                    disabled={!selectedStaff}
                    onClick={(e) => {
                        e.preventDefault();
                        handleDeleteUser();
                        setRefresh(true);
                    }}
                >
                    Delete Report
                </Button>
                <LinkButton
                    className={styles.button}
                    name="Add Skill"
                    path={`/addstaffskill/${selectedStaff?.report?.id || ""}`}
                />
                <LinkButton
                    className={styles.button}
                    name="Edit Skill"
                    disabled={!selectedStaffSkill}
                    path={`/edit/staffskill/${selectedStaffSkill?.id || ""}`}
                />
                <Button
                    className={styles.button}
                    disabled={!selectedStaffSkill}
                    onClick={(e) => {
                        e.preventDefault();
                        handleDeleteStaffSkill();
                        setRefresh(true);
                    }}
                >
                    Delete Skill
                </Button>
            </div>
        </form>
    );
};
