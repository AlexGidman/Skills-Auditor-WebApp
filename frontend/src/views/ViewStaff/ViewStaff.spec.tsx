import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { mockAdminUser, mockApiRequests, mockError } from "../../setupTests";
import * as ReactDom from "react-router-dom";
import { ViewStaff } from "./ViewStaff";
import { wait } from "@testing-library/user-event/dist/utils";
import React from "react";
import { Category, DirectReport, Skill, User } from "../../utility/types";

jest.mock("react-router-dom", () => {
    return {
        ...jest.requireActual("react-router-dom"),
        useOutletContext: jest.fn(),
        useNavigate: jest.fn(),
    };
});

const mockUser1: User = {
    id: "1",
    firstName: "testfirstName1",
    lastName: "testlastName1",
    email: "test1@email.com",
    password: "testPassword",
    systemRole: "Manager",
    jobRole: "Mid-Level Developer",
};

const mockUser2: User = {
    id: "2",
    firstName: "testfirstName2",
    lastName: "testlastName2",
    email: "test2@email.com",
    password: "testPassword",
    systemRole: "StaffUser",
    jobRole: "Senior Developer",
};

const mockCategory: Category = { id: "1", name: "testname1" };

const mockSkill1: Skill = {
    id: "1",
    name: "Skill1",
    categoryId: mockCategory.id,
    category: mockCategory,
};

const mockSkill2: Skill = {
    id: "2",
    name: "Skill2",
    categoryId: mockCategory.id,
    category: mockCategory,
};

const mockReports: DirectReport[] = [
    { id: "10", report: mockUser1 },
    { id: "20", report: mockUser2 },
];

describe("ViewStaff", () => {
    beforeEach(() => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue([mockAdminUser, jest.fn()]);
        jest.spyOn(ReactDom, "useNavigate").mockImplementation(() => jest.fn());
    });

    it.only("renders correctly when data retrieval successful", async () => {
        // @ts-ignore TODO: fix this
        mockApiRequests.getAllDirectReports.mockImplementation(() =>
            Promise.resolve({ data: mockReports }),
        );
        // @ts-ignore TODO: fix this
        mockApiRequests.getAllStaffSkills.mockImplementation(() =>
            Promise.resolve({ data: [{ id: "100", skill: mockSkill1, skillLevel: "1" }] }),
        );
        const { container } = render(
            <ReactDom.MemoryRouter>
                <ViewStaff />
            </ReactDom.MemoryRouter>,
        );

        expect(await screen.findByRole("heading", { name: "View Staff" })).toBeInTheDocument();

        expect(await screen.findByRole("listbox", { name: "Direct Reports" })).toBeInTheDocument();
        expect(
            await screen.findByRole("option", {
                name: `${mockUser1.firstName} ${mockUser1.lastName}`,
            }),
        ).toBeInTheDocument();
        expect(
            await screen.findByRole("option", {
                name: `${mockUser2.firstName} ${mockUser2.lastName}`,
            }),
        ).toBeInTheDocument();

        expect(screen.getByRole("table")).toBeInTheDocument();
        expect(screen.getByRole("columnheader", { name: "Email" })).toBeInTheDocument();
        expect(screen.getByRole("cell", { name: mockUser1.email })).toBeInTheDocument();
        expect(screen.getByRole("columnheader", { name: "Job role" })).toBeInTheDocument();
        expect(screen.getByRole("cell", { name: mockUser1.jobRole })).toBeInTheDocument();
        expect(screen.getByRole("columnheader", { name: "System role" })).toBeInTheDocument();
        expect(screen.getByRole("cell", { name: mockUser1.systemRole })).toBeInTheDocument();

        expect(await screen.findByRole("listbox", { name: "Skills" })).toBeInTheDocument();

        expect(
            await screen.findByRole("option", { name: `${mockSkill1.name} - None` }),
        ).toBeInTheDocument();

        expect(screen.getByRole("link", { name: "Edit User" })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: "Delete Report" })).toBeInTheDocument();
        expect(screen.getByRole("link", { name: "Add Skill" })).toBeInTheDocument();
        expect(screen.getByRole("link", { name: "Edit Skill" })).toBeInTheDocument();
        await wait(1000);

        expect(screen.getByRole("link", { name: "Edit User" })).toHaveAttribute(
            "href",
            `/edit/user/${mockUser1.id}`,
        );
        expect(screen.getByRole("link", { name: "Add Skill" })).toHaveAttribute(
            "href",
            `/addstaffskill/${mockAdminUser.id}`,
        );
        expect(screen.getByRole("link", { name: "Edit Skill" })).toHaveAttribute(
            "href",
            `/edit/staffskill/100`,
        );

        expect(container).toMatchSnapshot();
    });

    it("renders user info and links correctly on Direct Reports listbox option selection", async () => {
        // @ts-ignore TODO: fix this
        mockApiRequests.getAllDirectReports.mockImplementation(() =>
            Promise.resolve({ data: mockReports }),
        );
        // @ts-ignore TODO: fix this
        mockApiRequests.getAllStaffSkills.mockImplementation(() =>
            Promise.resolve({ data: [{ id: "100", skill: mockSkill1, skillLevel: "1" }] }),
        );
        render(
            <ReactDom.MemoryRouter>
                <ViewStaff />
            </ReactDom.MemoryRouter>,
        );

        expect(await screen.findByRole("heading", { name: "View Staff" })).toBeInTheDocument();

        expect(await screen.findByRole("listbox", { name: "Direct Reports" })).toBeInTheDocument();
        expect(
            await screen.findByRole("option", {
                name: `${mockUser1.firstName} ${mockUser1.lastName}`,
            }),
        ).toBeInTheDocument();
        expect(
            await screen.findByRole("option", {
                name: `${mockUser2.firstName} ${mockUser2.lastName}`,
            }),
        ).toBeInTheDocument();

        expect(screen.getByRole("table")).toBeInTheDocument();
        expect(screen.getByRole("columnheader", { name: "Email" })).toBeInTheDocument();
        expect(screen.getByRole("cell", { name: mockUser1.email })).toBeInTheDocument();
        expect(screen.getByRole("columnheader", { name: "Job role" })).toBeInTheDocument();
        expect(screen.getByRole("cell", { name: mockUser1.jobRole })).toBeInTheDocument();
        expect(screen.getByRole("columnheader", { name: "System role" })).toBeInTheDocument();
        expect(screen.getByRole("cell", { name: mockUser1.systemRole })).toBeInTheDocument();
        expect(screen.getByRole("link", { name: "Edit User" })).toHaveAttribute(
            "href",
            `/edit/user/${mockUser1.id}`,
        );

        userEvent.selectOptions(
            await screen.findByRole("listbox", { name: "Direct Reports" }),
            mockUser2.id,
        );
        expect(screen.queryByRole("cell", { name: mockUser1.email })).not.toBeInTheDocument();
        expect(screen.getByRole("cell", { name: mockUser2.email })).toBeInTheDocument();
        expect(screen.queryByRole("cell", { name: mockUser1.jobRole })).not.toBeInTheDocument();
        expect(screen.getByRole("cell", { name: mockUser2.jobRole })).toBeInTheDocument();
        expect(screen.queryByRole("cell", { name: mockUser1.systemRole })).not.toBeInTheDocument();
        expect(screen.getByRole("cell", { name: mockUser2.systemRole })).toBeInTheDocument();
        expect(screen.getByRole("link", { name: "Edit User" })).toHaveAttribute(
            "href",
            `/edit/user/${mockUser2.id}`,
        );
    });

    it("renders edit staff skill link correctly on Skills listbox option selection", async () => {
        // @ts-ignore TODO: fix this
        mockApiRequests.getAllDirectReports.mockImplementation(() =>
            Promise.resolve({ data: mockReports }),
        );
        // @ts-ignore TODO: fix this
        mockApiRequests.getAllStaffSkills.mockImplementation(() =>
            Promise.resolve({
                data: [
                    { id: "100", skill: mockSkill1, skillLevel: "1" },
                    { id: "200", skill: mockSkill2, skillLevel: "2" },
                ],
            }),
        );
        render(
            <ReactDom.MemoryRouter>
                <ViewStaff />
            </ReactDom.MemoryRouter>,
        );

        expect(await screen.findByRole("heading", { name: "View Staff" })).toBeInTheDocument();

        expect(await screen.findByRole("listbox", { name: "Skills" })).toBeInTheDocument();
        expect(
            await screen.findByRole("option", { name: `${mockSkill1.name} - None` }),
        ).toBeInTheDocument();
        expect(
            await screen.findByRole("option", { name: `${mockSkill2.name} - Basic` }),
        ).toBeInTheDocument();
        await wait(100);

        expect(screen.getByRole("link", { name: "Edit Skill" })).toHaveAttribute(
            "href",
            `/edit/staffskill/100`,
        );

        userEvent.selectOptions(await screen.findByRole("listbox", { name: "Skills" }), "200");
        await wait();
        expect(screen.getByRole("link", { name: "Edit Skill" })).toHaveAttribute(
            "href",
            `/edit/staffskill/200`,
        );
    });

    it("should have delete report button that refreshes the page following deletion of user", async () => {
        // @ts-ignore TODO: fix this
        mockApiRequests.getAllDirectReports.mockImplementation(() =>
            Promise.resolve({ data: mockReports }),
        );
        // @ts-ignore TODO: fix this
        mockApiRequests.getAllStaffSkills.mockImplementation(() =>
            Promise.resolve({
                data: [
                    { id: "100", skill: mockSkill1, skillLevel: "1" },
                    { id: "200", skill: mockSkill2, skillLevel: "2" },
                ],
            }),
        );
        // @ts-ignore TODO: fix this
        mockApiRequests.deleteUser.mockImplementation(() => Promise.resolve({ data: "Success" }));
        const mockNavigate = jest.fn();
        jest.spyOn(ReactDom, "useNavigate").mockImplementation(() => mockNavigate);

        render(
            <ReactDom.MemoryRouter>
                <ViewStaff />
            </ReactDom.MemoryRouter>,
        );

        expect(await screen.findByRole("heading", { name: "View Staff" })).toBeInTheDocument();

        expect(await screen.findByRole("listbox", { name: "Direct Reports" })).toBeInTheDocument();
        expect(
            await screen.findByRole("option", {
                name: `${mockUser1.firstName} ${mockUser1.lastName}`,
            }),
        ).toBeInTheDocument();
        expect(
            await screen.findByRole("option", {
                name: `${mockUser2.firstName} ${mockUser2.lastName}`,
            }),
        ).toBeInTheDocument();

        expect(mockNavigate).not.toHaveBeenCalled();
        userEvent.click(screen.getByRole("button", { name: "Delete Report" }));
        await wait(100);
        expect(mockNavigate).toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenLastCalledWith(0);
    });

    it("renders correctly when data retrieval fails", async () => {
        mockApiRequests.getAllDirectReports.mockImplementation(() => Promise.reject(mockError));

        const { container } = render(
            <ReactDom.MemoryRouter>
                <ViewStaff />
            </ReactDom.MemoryRouter>,
        );

        expect(await screen.findByRole("alert")).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });
});
