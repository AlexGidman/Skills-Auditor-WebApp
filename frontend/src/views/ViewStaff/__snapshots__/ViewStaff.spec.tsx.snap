// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewStaff renders correctly when data retrieval fails 1`] = `
<div>
  <div
    aria-label="Oops"
    class="component"
    role="alert"
  >
    <h1>
      Something went wrong...
    </h1>
    <p>
      Oops
    </p>
  </div>
</div>
`;

exports[`ViewStaff renders correctly when data retrieval successful 1`] = `
<div>
  <header>
    <h1>
      View Staff
    </h1>
  </header>
  <section>
    <form
      class="form"
    >
      <div
        class="component component listview"
      >
        <label
          class="label"
          for="Direct-Reports-undefined"
        >
          Direct Reports
          :
        </label>
        <select
          aria-label="Direct Reports"
          class="select"
          name="Direct Reports"
          size="5"
        >
          <option
            value="1"
          >
            testfirstName1 testlastName1
          </option>
          <option
            value="2"
          >
            testfirstName2 testlastName2
          </option>
        </select>
      </div>
      <div
        class="flexbox"
      >
        <table
          class="table"
        >
          <tbody>
            <tr>
              <th>
                Email
              </th>
              <td>
                test1@email.com
              </td>
            </tr>
            <tr>
              <th>
                Job role
              </th>
              <td>
                Mid-Level Developer
              </td>
            </tr>
            <tr>
              <th>
                System role
              </th>
              <td>
                Manager
              </td>
            </tr>
          </tbody>
        </table>
        <div
          class="component component skillsList"
        >
          <label
            class="label"
            for="Skills-undefined"
          >
            Skills
            :
          </label>
          <select
            aria-label="Skills"
            class="select"
            name="Skills"
            size="5"
          >
            <option
              value="100"
            >
              Skill1 - None
            </option>
          </select>
        </div>
      </div>
      <div
        class="flexbox"
      >
        <a
          class="component button"
          href="/edit/user/1"
        >
          Edit User
        </a>
        <button
          class="component button"
        >
          Delete Report
        </button>
        <a
          class="component button"
          href="/addstaffskill/1"
        >
          Add Skill
        </a>
        <a
          class="component button"
          href="/edit/staffskill/100"
        >
          Edit Skill
        </a>
        <button
          class="component button"
        >
          Delete Skill
        </button>
      </div>
    </form>
  </section>
</div>
`;
